crates/ty/CONTRIBUTING.md

```diff
@@ -0,0 +1,127 @@
+when making a commit:
+
+```shell
+uv tool install pre-commit
+pre-commit install
+```
+
+We recommend [nextest](https://nexte.st/) to run Ruff's test suite (via `cargo nextest run`),
+though it's not strictly necessary:
+
+```shell
+cargo install cargo-nextest --locked
+```
+
+Throughout this guide, any usages of `cargo test` can be replaced with `cargo nextest run`,
+if you choose to install `nextest`.
+
+### Development
+
+After cloning the repository, run ty locally from the repository root with:
+
+```shell
+cargo run --bin ty -- check --project /path/to/project/
+```
+
+Prior to opening a pull request, ensure that your code has been auto-formatted,
+and that it passes both the lint and test validation checks:
+
+```shell
+cargo clippy --workspace --all-targets --all-features -- -D warnings  # Rust linting
+cargo test  # Rust testing
+uvx pre-commit run --all-files --show-diff-on-failure  # Rust and Python formatting, Markdown and Python linting, etc.
+```
+
+These checks will run on GitHub Actions when you open your pull request, but running them locally
+will save you time and expedite the merge process.
+
+If you're using VS Code, you can also install the recommended [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer) extension to get these checks while editing.
+
+Include the text `[ty]` at the beginning of your pull request title, to distinguish ty pull requests
+from Ruff ones.
+
+Your pull request will be reviewed by a maintainer, which may involve a few rounds of iteration
+prior to merging.
+
+### Project Structure
+
+Ruff is structured as a monorepo with a [flat crate structure](https://matklad.github.io/2021/08/22/large-rust-workspaces.html),
+such that all crates are contained in a flat `crates` directory.
+
+The vast majority of the code lives in the `ty_python_semantic` crate (located at
+`crates/ty_python_semantic`). As a contributor, that's the crate that'll probably be most relevant
+to you.
+
+At the time of writing, the repository includes the following ty-specific crates (in addition to
+crates shared with Ruff, such as `ruff_db`, `ruff_python_ast`, and `ruff_python_parser`):
+
+- `ty_python_semantic`: The core type checker, which includes the type inference engine and
+    semantic analysis.
+- `ty_test`: The Markdown-based test framework for ty, "mdtest".
+- `ty`: The command-line interface.
+- `ty_ide`: IDE features (hover, go-to-definition, autocomplete) for the language server.
+- `ty_project`: Discovery and representation of a Python project to be checked by ty.
+- `ty_server`: The ty language server.
+- `ty_vendored`: A vendored copy of [typeshed](https://github.com/python/typeshed), which holds type
+    annotations for the Python standard library.
+- `ty_wasm`: library crate for exposing ty as a WebAssembly module. Powers the
+    [Ty Playground](https://play.ty.dev/).
+
+## Writing tests
```

Comment:
I think we should also add a brief mention of property tests and either link to https://github.com/astral-sh/ruff/blob/bdccb37b4ae17606015d4c03e04f1dd6f1d6f4da/crates/ty_python_semantic/src/types/property_tests.rs#L1-L25 or extract the content into markdown format and link to that instead.